File: src/main/java/com/github/joelws/release/tracker/util/JsonAdapterImpl.java
package com.github.joelws.release.tracker.util;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.joelws.release.tracker.exception.ReleaseTrackerException;
import com.github.joelws.release.tracker.handler.JsonResponse;
import com.github.joelws.release.tracker.interfaces.JsonAdapter;

import org.apache.log4j.Logger;

import java.io.IOException;

public class JsonAdapterImpl implements JsonAdapter {
    private static final Logger LOGGER = Logger.getLogger(JsonAdapterImpl.class);

    @Override
    public String getJsonFromObject(Object obj) {
        try {
            final ObjectMapper mapper = new ObjectMapper();
            return mapper.writeValueAsString(obj);
        } catch (Exception e) {
            LOGGER.info("Unable to map object to JSON");
            throw new ReleaseTrackerException(new JsonResponse(400, "Invalid format"));
        }
    }

    @Override
    public <T> T getObjectFromJson(String jsonString, Class<T> klazz) {
        try {
            final ObjectMapper mapper = new ObjectMapper();
            final T obj = mapper.readValue(jsonString, klazz);
            return obj;

        } catch (IOException e) {
            LOGGER.info("Unable to map JSON to object");
            throw new ReleaseTrackerException(new JsonResponse(400, "Invalid format"));
        }
    }
}

File: src/main/kotlin/com/github/joelws/release/tracker/util/JsonAdapterImpl.kt
package com.github.joelws.release.tracker.util

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.github.joelws.release.tracker.exception.ReleaseTrackerException
import com.github.joelws.release.tracker.interfaces.JsonAdapter
import com.github.joelws.release.tracker.response.RestResponse.BadRequest
import org.apache.log4j.Logger
import java.io.IOException

open class JsonAdapterImpl : JsonAdapter {

    companion object {
        private val LOGGER = Logger.getLogger(JsonAdapterImpl::class.java)
    }

    override fun getJsonFromObject(any: Any): String? = try {
        jacksonObjectMapper().writeValueAsString(any)
    } catch (e: Exception) {
        LOGGER.info("Unable to map object to JSON: $any")
        throw ReleaseTrackerException(BadRequest())
    }


    override fun <T> getObjectFromJson(jsonString: String?, klazz: Class<T>): T = try {
        jacksonObjectMapper().readValue(jsonString, klazz)
    } catch (e: IOException) {
        LOGGER.info("Unable to map JSON to object: $jsonString")
        throw ReleaseTrackerException(BadRequest())
    }
}
